File: kotlin-node/karakum/injections/FsStatsMembersInjection.js
import ts from "typescript";
import * as karakum from "karakum";

export default {
    setup(context) {
        this.statsBaseNodes = new Map()
    },

    traverse(node) {
        const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

        if (
            sourceFileName.endsWith("fs.d.ts")
            && ts.isMethodSignature(node)
            && ts.isIdentifier(node.name)

            && node.parent
            && ts.isInterfaceDeclaration(node.parent)
            && node.parent.name.text === "StatsBase"
        ) {
            this.statsBaseNodes.set(node.name.text + node.parameters.length, node)
        }

        if (
            sourceFileName.endsWith("fs.d.ts")
            && ts.isPropertySignature(node)
            && ts.isIdentifier(node.name)

            && node.parent
            && ts.isInterfaceDeclaration(node.parent)
            && node.parent.name.text === "StatsBase"
        ) {
            this.statsBaseNodes.set(node.name.text, node)
        }
    },

    render(node, context, next) {
        return null
    },

    inject(node, context, render) {
        const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

        if (
            sourceFileName.endsWith("fs.d.ts")
            && ts.isClassDeclaration(node)
            && node?.name.text === "Stats"
            && !context.static
        ) {
            return Array.from(this.statsBaseNodes.values())
                .map(member => {
                    if (ts.isPropertySignature(member)) {
                        const readonly = member.modifiers?.find(modifier => modifier.kind === ts.SyntaxKind.ReadonlyKeyword)

                        const modifier = readonly
                            ? "val "
                            : "var "

                        const name = karakum.escapeIdentifier(render(member.name))

                        const isOptional = Boolean(member.questionToken)

                        let type = karakum.renderNullable(member.type, isOptional, context, render)

                        if (type === "T") {
                            type = "Double"
                        }

                        return `override ${modifier}${name}: ${type}`
                    }

                    const name = karakum.escapeIdentifier(render(member.name))

                    const typeParameters = member.typeParameters
                        ?.map(typeParameter => render(typeParameter))
                        ?.join(", ")

                    const returnType = member.type && render(member.type)

                    return karakum.convertParameterDeclarations(member, context, render, {
                        strategy: "function",
                        defaultValue: "",
                        template: parameters => {
                            return `override fun ${karakum.ifPresent(typeParameters, it => `<${it}> `)}${name}(${parameters})${karakum.ifPresent(returnType, it => `: ${it}`)}`
                        }
                    })
                })
                .join("\n")
        }

        return null
    },

    generate(context) {
        return []
    }
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/injections/FsStatsMembersInjection.kt
package node.karakum.injections

import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.GeneratedFile
import io.github.sgrishchenko.karakum.extension.Injection
import io.github.sgrishchenko.karakum.extension.InjectionContext
import io.github.sgrishchenko.karakum.extension.InjectionType
import io.github.sgrishchenko.karakum.extension.Render
import io.github.sgrishchenko.karakum.extension.ifPresent
import io.github.sgrishchenko.karakum.extension.plugins.ParameterDeclarationStrategy
import io.github.sgrishchenko.karakum.extension.plugins.ParameterDeclarationsConfiguration
import io.github.sgrishchenko.karakum.extension.plugins.convertParameterDeclarations
import io.github.sgrishchenko.karakum.extension.renderNullable
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import arrow.core.raise.impure
import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.MEMBER
import io.github.sgrishchenko.karakum.extension.plugins.function
import typescript.Node
import typescript.SyntaxKind
import typescript.asArray
import typescript.isClassDeclaration
import typescript.isInterfaceDeclaration
import typescript.isMethodSignature
import typescript.isPropertySignature

class FsStatsMembersInjection : Injection {
    private val statsBaseNodes = mutableListOf<Node>()

    override fun setup(context: Context) = Unit

    override fun traverse(node: Node, context: Context) = impure {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("fs.d.ts"))

        nullable {
            ensure(isMethodSignature(node))

            val interfaceNode = ensureNotNull(node.getParentOrNull())
            ensure(isInterfaceDeclaration(interfaceNode))
            ensure(interfaceNode.name.text == "StatsBase")

            statsBaseNodes += node
        } ?: nullable {
            ensure(isPropertySignature(node))

            val interfaceNode = ensureNotNull(node.getParentOrNull())
            ensure(isInterfaceDeclaration(interfaceNode))
            ensure(interfaceNode.name.text == "StatsBase")

            statsBaseNodes += node
        }
    }

    override fun render(node: Node, context: Context, next: Render<Node>) = null

    override fun inject(node: Node, context: InjectionContext, render: Render<Node>) = nullable {
        ensure(context.type == InjectionType.MEMBER)

        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("fs.d.ts"))

        ensure(isClassDeclaration(node))
        ensure(node.name?.text == "Stats")

        statsBaseNodes
            .mapNotNull { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val readonly = member.modifiers?.asArray()?.find { it.kind == SyntaxKind.ReadonlyKeyword }

                    val modifier = if (readonly != null) "val " else "var "

                    val name = render(member.name)

                    val isOptional = member.questionToken != null

                    var type = renderNullable(member.type, isOptional, context, render)

                    if (type == "T") {
                        type = "Double"
                    }

                    "override ${modifier}${name}: $type"
                } ?: nullable {
                    ensure(isMethodSignature(member))

                    val name = render(member.name)

                    val typeParameters = member.typeParameters?.asArray()
                        ?.joinToString(", ") { render(it) }

                    val returnType = member.type?.let { render(it) }

                    convertParameterDeclarations(
                        member, context, render,
                        ParameterDeclarationsConfiguration(
                            strategy = ParameterDeclarationStrategy.function,
                            defaultValue = "",
                            template = { parameters, _ ->
                                "override fun ${ifPresent(typeParameters) { "<${it}> " }}${name}(${parameters})${ifPresent(returnType) { ": $it" }}"
                            }
                        )
                    )
                }
            }
            .toTypedArray()
    }

    override fun generate(context: Context, render: Render<Node>) = emptyArray<GeneratedFile>()
}
